### PR TITLE
feat: Complete v0.4.0 milestone — 13 new features, 4 benchmarks, 622 tests

### LINKED ISSUE #64: [Feature]: Custom Metrics Framework
## Problem Statement
Users need to define custom metrics beyond standard accuracy/pass rates.

## Proposed Solution
Implement custom metrics framework:
- Define metric plugin interface
- Support composite metrics
- Add metric aggregation strategies
- Allow custom scoring functions
- Provide metric validation

## Feature Area
Analytics - Extensibility

## Checklist
- [ ] Define metric interface
- [ ] Implement metric registry
- [ ] Add aggregation strategies
- [ ] Support custom scoring
- [ ] Add metric validation
- [ ] Create example metrics
- [ ] Write unit tests
- [ ] Update documentation

🤖 Generated from roadmap

### LINKED ISSUE #129: [Feature]: Latency and Performance Benchmarks
## Problem Statement
Need to measure and track response latency and throughput.

## Proposed Solution
Implement performance benchmarks:
- Measure time-to-first-token
- Track total response time
- Measure throughput (tokens/sec)
- Add percentile calculations (p50, p95, p99)
- Support load testing

## Feature Area
Benchmarks - Performance

## Checklist
- [ ] Add latency measurement
- [ ] Add throughput measurement
- [ ] Implement percentile calculations
- [ ] Add load testing support
- [ ] Create performance reports
- [ ] Write unit tests
- [ ] Update documentation

🤖 Generated from roadmap

### LINKED ISSUE #138: [Feature]: Dataset Versioning
## Problem Statement
Need to track and manage different versions of benchmark datasets.

## Proposed Solution
Implement dataset versioning:
- Add version tracking
- Support multiple dataset versions
- Add dataset changelog
- Implement version pinning
- Add migration tools

## Feature Area
Benchmarks - Data Management

## Checklist
- [ ] Design versioning schema
- [ ] Add version tracking
- [ ] Support multiple versions
- [ ] Add changelog
- [ ] Add version pinning
- [ ] Create migration tools
- [ ] Update documentation

🤖 Generated from roadmap

### LINKED ISSUE #142: [Feature]: Sample Benchmarks
## Problem Statement
Need to support dataset sampling for faster iteration during development.

## Proposed Solution
Implement sampling features:
- Add random sampling
- Support stratified sampling
- Add configurable sample sizes
- Support sampling strategies
- Maintain sample reproducibility

## Feature Area
Benchmarks - Development

## Checklist
- [ ] Add random sampling
- [ ] Add stratified sampling
- [ ] Add size configuration
- [ ] Add sampling strategies
- [ ] Add seed control
- [ ] Write unit tests
- [ ] Update documentation

🤖 Generated from roadmap

### LINKED ISSUE #232: Add integration tests for MCPToolBench++ benchmark
## Overview
Add integration tests for MCPToolBench++ benchmark implementation to ensure quality and prevent regressions.

## Background
Depends on MCPToolBench++ integration (#223). Once implemented, need comprehensive tests to validate functionality.

### Unit Tests
```python
# In tests/test_benchmarks.py
class TestMCPToolBenchmark:
    """Unit tests for MCPToolBench++ benchmark."""
    
    def test_load_tasks(self):
        """Test loading tasks from HuggingFace dataset."""
        benchmark = create_benchmark("mcp-toolbench")
        tasks = benchmark.load_tasks(sample_size=1)
        assert len(tasks) == 1
    
    def test_load_specific_tasks(self):
        """Test loading specific task IDs."""
        benchmark = create_benchmark("mcp-toolbench")
        tasks = benchmark.load_tasks(task_ids=["task-123"])
        assert len(tasks) == 1
        assert tasks[0]["task_id"] == "task-123"
    
    def test_normalize_task(self):
        """Test task normalization to BenchmarkTask."""
        benchmark = create_benchmark("mcp-toolbench")
        task = {...}  # Mock task data
        normalized = benchmark.normalize_task(task)
        assert isinstance(normalized, BenchmarkTask)
        assert normalized.task_id == task["task_id"]
    
    def test_prompt_template(self):
        """Test prompt template generation."""
        benchmark = create_benchmark("mcp-toolbench")
        template = benchmark.get_prompt_template()
        assert "{problem_statement}" in template
    
    def test_no_prebuilt_images(self):
        """Test that MCPToolBench has no pre-built images."""
        benchmark = create_benchmark("mcp-toolbench")
        task = {...}
        assert benchmark.get_prebuilt_image(task) is None
```

### Integration Tests
```python
# In tests/test_integration.py
@pytest.mark.integration
async def test_mcptoolbench_full_workflow(mock_mcp_server):
    """Test complete MCPToolBench workflow."""
    config = create_test_config(benchmark="mcp-toolbench")
    
    # Run benchmark on single task
    results = await run_benchmark(
        config=config,
        sample_size=1,
        run_mcp=True,
        run_baseline=False,
    )
    
    assert len(results["tasks"]) == 1
    assert "resolved" in results["tasks"][0]["mcp"]
    assert "tool_usage" in results["tasks"][0]["mcp"]

@pytest.mark.integration
async def test_mcptoolbench_evaluation():
    """Test MCPToolBench evaluation logic."""
    benchmark = create_benchmark("mcp-toolbench")
    
    # Create mock environment
    task = {...}
    env = await create_mock_environment()
    
    # Test evaluation
    solution = "..."  # Mock solution
    result = await benchmark.evaluate(env, task, solution)
    
    assert "resolved" in result
    assert isinstance(result["resolved"], bool)
```

### CLI Tests
```python
# In tests/test_cli.py
def test_cli_mcptoolbench(runner):
    """Test MCPToolBench via CLI."""
    result = runner.invoke(cli, [
        "run",
        "-c", "config.yaml",
        "--benchmark", "mcp-toolbench",
        "-n", "1",
    ])
    assert result.exit_code == 0
    assert "mcp-toolbench" in result.output

def test_cli_list_benchmarks():
    """Test that MCPToolBench appears in benchmarks list."""
    result = runner.invoke(cli, ["benchmarks"])
    assert "mcp-toolbench" in result.output
```

## Implementation Tasks
- [ ] Add unit tests in `tests/test_benchmarks.py`
- [ ] Add integration tests in `tests/test_integration.py`
- [ ] Add CLI tests in `tests/test_cli.py`
- [ ] Create mock MCPToolBench tasks for testing
- [ ] Test error handling and edge cases
- [ ] Add tests for tool use evaluation metrics
- [ ] Ensure tests run in CI/CD pipeline

## Acceptance Criteria
- [ ] All test categories implemented (unit, integration, CLI)
- [ ] Tests cover happy path and error cases
- [ ] Mock data realistic and representative
- [ ] Tests pass consistently in CI
- [ ] Code coverage >80% for MCPToolBench code
- [ ] Tests document expected behavior

## Dependencies
- Blocked by MCPToolBench++ integration (#223)

## Roadmap
Phase 1: Foundation (v0.3.0)

## Effort Estimate
Small (1-2 days)

## Impact
Medium - Ensures quality as feature develops, prevents regressions

### PR DESCRIPTION
**Pull request by greynewell**
## Summary

Completes the v0.4.0 Benchmarks milestone by implementing all 13 remaining open issues. This adds 4 new benchmark implementations, 7 new framework modules, and comprehensive test coverage (622 new tests, 1809 total).

## New Benchmarks

- **Custom benchmarks via YAML** (#29, #47): Define benchmarks without writing Python — supports exact_match, numeric, regex, and script evaluation types
- **MMMU multi-modal benchmark** (#123): Massive Multi-discipline Multimodal Understanding for image understanding tasks
- **LongBench long-context benchmark** (#125): 21 subsets across 6 categories with F1, ROUGE-L, classification accuracy, and edit similarity metrics
- **Adversarial testing benchmark** (#126): Safety and robustness testing using HarmBench with refusal detection across jailbreak, hallucination, bias, and robustness categories

## New Framework Modules

- **Custom metrics** (#64): Define and compute custom evaluation metrics with composite metrics support and built-in registry (7 default metrics)
- **Failure analysis** (#67): Categorize failures into 9 categories with pattern extraction and actionable recommendations
- **Sampling strategies** (#142): Sequential, random, and stratified sampling with seed-based reproducibility
- **Dataset versioning** (#138): Pin HuggingFace dataset versions with manifest save/load for reproducible runs
- **Latency metrics** (#129): Track task latency, time-to-first-tool-call, throughput, and percentile statistics (p50/p95/p99)
- **GPU support** (#121): NVIDIA GPU detection and Docker container GPU configuration
- **Few-shot learning** (#127): Variable shot counts with random/similar/diverse selection strategies and learning curve analysis

## Also Includes

- **MCPToolBench++ integration tests** (#232): 60 tests for the existing MCPToolBench++ benchmark
- Updated README benchmark table (25+ → 30+ benchmarks across 10 categories)
- Updated roadmap Phase 2 to reflect completion
- Example custom benchmark YAML configuration

## Test Coverage

All 1809 tests pass (622 new). Coverage on new modules:

| Module | Coverage |
|--------|----------|
| adversarial.py | 100% |
| custom.py | 96% |
| longbench.py | 98% |
| mmmu.py | 94% |
| custom_metrics.py | 100% |
| failure_analysis.py | 99% |
| sampling.py | 99% |
| dataset_versioning.py | 100% |
| latency_metrics.py | 100% |
| gpu_support.py | 94% |
| few_shot.py | 98% |

## Test plan

- [x] `uvx ruff check src/ tests/` — All checks passed
- [x] `uvx ruff format --check src/ tests/` — All formatted
- [x] `uv run pytest -m "not integration"` — 1809 passed, 0 failures
- [x] Pre-commit hooks pass
- [ ] CI pipeline passes (lint, format, build, tests on Python 3.11 + 3.12)

Closes #29, closes #47, closes #64, closes #67, closes #121, closes #123, closes #125, closes #126, closes #127, closes #129, closes #138, closes #142, closes #232

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Custom benchmarks via YAML definitions
  * Custom metrics and failure analysis capabilities
  * Random and stratified sampling strategies
  * Dataset versioning with manifest support
  * Latency and performance metrics (task latency, time-to-first-tool-call, throughput percentiles)
  * GPU support for Docker containers
  * Few-shot learning support with multiple selection strategies
  * Four new benchmarks: MMMU (multimodal), LongBench (long-context), Adversarial (safety), and MCPToolBench++ integration tests
  * Expanded to 30+ benchmarks across 10 categories

* **Documentation**
  * Example custom benchmark YAML configuration

* **Tests**
  * Comprehensive test coverage for all new modules

### CODE REVIEW

**Comment by greynewell** [comment]
@coderabbitai review

File: src/mcpbr/benchmarks/custom.py [hunk 1]
```diff
@@ -0,0 +1,606 @@
...
+
+        # Write solution and ground truth to temp files in the container
+        escaped_solution = solution.replace("'", "'\\''")
+        escaped_truth = ground_truth.replace("'", "'\\''")
+
+        setup_cmd = (
+            f"echo '{escaped_solution}' > /tmp/solution.txt && "
+            f"echo '{escaped_truth}' > /tmp/ground_truth.txt"
+        )
+        await env.exec_command(setup_cmd, timeout=30)
```
[the remark this answers is not in the document]
Comment (greynewell) [reply src/mcpbr/benchmarks/custom.py]: ✅ **Resolved in commit 6b9f373**

Fixed by switching to base64 encoding as suggested:
- Solution and ground truth are now base64-encoded in Python before being passed to the shell
- The shell command uses `echo '<b64>' | base64 -d > /tmp/file.txt` instead of direct string interpolation
- This completely eliminates shell injection risk from `$(...)`, backticks, or any other metacharacters

All 72 custom benchmark tests continue to pass.
